### ISSUE TITLE
Float16 model

### ISSUE THREAD

**Issue by Osbornnnnn** [yalue/onnxruntime_go#20]
If I want to run the model on float16, I get ...

`panic: Error running network: Unexpected input data type. Actual: (tensor(uint16)) , expected: (tensor(float16))`

**Comment by yalue** [comment]
Hi, and thanks for pointing this out.  The reason is that the library internally detects the type of the Tensor[uint16] and tells onnxruntime that it's a uint16.  To solve this problem (since go doesn't support 16-bit floats), I'll need to add the ability for users to manually create tensors backed by arbitrary data.  This probably won't be that hard to add, fortunately.  I don't have any timeline estimates yet, but I can work on this next time I have a chance.  Probably some time in the next week or two.

**Comment by Osbornnnnn** [comment]
THX my lover

**Comment by yalue** [comment]
Alright, this should be available as of commit 42f85c9f58692e2df0df13c7f14618f4558b9eac.  To create a float16 tensor, you'll need to convert your float16 data into an array of bytes.  As of the latest commit (version `v1.3.0`), you can create float16 or bfloat16 tensors as follows:

```go
// CustomDataTensor instances are always backed by user-managed slices of bytes.
// The library will not try to infer a slice size or provide the GetData() function for these
// tensors, so the caller must create these slices.
var float16InputData []byte = getFloat16InputData()
var float16OutputData []byte := make([]byte, outputSizeInBytes)
inputTensor, err := ort.NewCustomDataTensor(inputShape, float16InputData, ort.TensorElementDataTypeFloat16)
if err != nil {
    // The underlying onnxruntime library should have caused an error here if the data
    // is insufficient for the size, if an invalid type was provided, etc.
}
defer inputTensor.Destroy()
outputTensor, err := ort.NewCustomDataTensor(outputShape, float16OutputData, ort.TensorElementDataTypeFloat16)
if err != nil {
    // ...
}
defer outputTensor.Destroy()
session, err := ort.NewSession("path/to/network.onnx",
    []string{"inputName"}, []string{"outputName"},
    []ArbitraryTensor{inputTensor}, []ArbitraryTensor{outputTensor},
    nil)
if err != nil {
    // ...
}
defer session.Destroy()

err = session.Run()
if err != nil {
    // ...
}

// At this point, the float16OutputData slice should contain the output data produced by the network.
```

Note that I haven't tested the example, but hopefully you get the gist.  The key point is that you can create a `CustomDataTensor` with any of the supported `TensorElementDataType` values.  (See the constants in the docs; they're just copied from the C header.)  After you've created a `CustomDataTensor`, you can use it with `Session` instances like the older `Tensor[T]` structs.  I hope this helps, and let me know if you encounter bugs with it!
